File: packages/dara-core/cypress/integration/loop_variable_nested.ts
import { loginBeforeRoute } from '../support/utils';

describe('LoopVariable in Variable.get() nested property', () => {
    beforeEach(loginBeforeRoute('/loop_variable_nested'));

    describe('Basic Dynamic Access', () => {
        it('displays values accessed via dynamic loop variable key', () => {
            cy.cardContent('Basic Dynamic Access').within(() => {
                // Verify all user names are displayed with correct values
                cy.contains('div', 'user1').parent().should('contain.text', 'Alice');
                cy.contains('div', 'user2').parent().should('contain.text', 'Bob');
            });
        });
    });

    describe('Editable Dynamic Access', () => {
        it('displays initial values correctly', () => {
            cy.cardContent('Editable Dynamic Access').within(() => {
                cy.get('input').eq(0).should('have.value', 'Value A');
                cy.get('input').eq(1).should('have.value', 'Value B');
            });
        });

        it('updates the underlying data when editing via dynamic key', () => {
            cy.cardContent('Editable Dynamic Access').within(() => {
                // Edit the first input (item_a)
                cy.get('input').eq(0).clear().type('New Value A');

                // Verify the data object is updated
                cy.contains('div', 'Current data object:')
                    .next()
                    .should('contain.text', 'New Value A')
                    .and('contain.text', 'Value B');
            });
        });

        it('edits do not affect other dynamically-accessed values', () => {
            cy.cardContent('Editable Dynamic Access').within(() => {
                // Edit the second input (item_b)
                cy.get('input').eq(1).clear().type('New Value B');

                // Verify only item_b changed
                cy.get('input').eq(0).should('have.value', 'Value A');
                cy.contains('div', 'Current data object:')
                    .next()
                    .should('contain.text', 'Value A')
                    .and('contain.text', 'New Value B');
            });
        });
    });

    describe('Deeply Nested Access', () => {
        it('displays values with mixed static and dynamic keys', () => {
            cy.cardContent('Deeply Nested Access').within(() => {
                // Each row displays: category / item : color
                // Use the unique item name to find the row, then verify the color
                cy.contains('div', 'apple').parent().should('contain.text', 'Red');
                cy.contains('div', 'carrot').parent().should('contain.text', 'Orange');
            });
        });
    });

    describe('Dynamic List Update', () => {
        it('displays initial single item', () => {
            cy.cardContent('Dynamic List Update').within(() => {
                // Use the unique value text to verify rows exist
                cy.contains('div', 'X Value').should('be.visible');
                // Should only have one item row - Y and Z values should not exist
                cy.contains('div', 'Y Value').should('not.exist');
                cy.contains('div', 'Z Value').should('not.exist');
            });
        });

        it('adds items and displays their dynamic values', () => {
            cy.cardContent('Dynamic List Update').within(() => {
                // Add second item (y)
                cy.get('button').contains('Add Item').click();
                cy.contains('div', 'Y Value').should('be.visible');

                // Add third item (z)
                cy.get('button').contains('Add Item').click();
                cy.contains('div', 'Z Value').should('be.visible');

                // All three should be visible
                cy.contains('div', 'X Value').should('be.visible');
                cy.contains('div', 'Y Value').should('be.visible');
                cy.contains('div', 'Z Value').should('be.visible');
            });
        });

        it('removes items correctly', () => {
            cy.cardContent('Dynamic List Update').within(() => {
                // Add all items first
                cy.get('button').contains('Add Item').click();
                cy.get('button').contains('Add Item').click();
                cy.contains('div', 'Z Value').should('be.visible');

                // Remove last item (z)
                cy.get('button').contains('Remove Item').click();
                cy.contains('div', 'Z Value').should('not.exist');
                cy.contains('div', 'Y Value').should('be.visible');

                // Remove another item (y)
                cy.get('button').contains('Remove Item').click();
                cy.contains('div', 'Y Value').should('not.exist');
                cy.contains('div', 'X Value').should('be.visible');
            });
        });

        it('shows placeholder when all items removed', () => {
            cy.cardContent('Dynamic List Update').within(() => {
                // Remove the initial item
                cy.get('button').contains('Remove Item').click();
                cy.contains('div', 'No items').should('be.visible');
            });
        });

        it('resets to initial state', () => {
            cy.cardContent('Dynamic List Update').within(() => {
                // Add items
                cy.get('button').contains('Add Item').click();
                cy.get('button').contains('Add Item').click();
                cy.contains('div', 'Z Value').should('be.visible');

                // Reset
                cy.get('button').contains('Reset').click();
                cy.contains('div', 'Y Value').should('not.exist');
                cy.contains('div', 'Z Value').should('not.exist');
                cy.contains('div', 'X Value').should('be.visible');
            });
        });
    });
});

File: packages/dara-core/cypress/dara/pages/loop_variable_nested.py
"""
LoopVariable in Variable.get() nested property E2E test page.

Tests using LoopVariable to dynamically access nested properties of a Variable inside a For loop.
"""

from dara.components import Button, Card, Input, Stack, Text
from dara.core import For, Variable, action
from dara.core.interactivity import ActionCtx


def loop_variable_nested():
    """LoopVariable nested property access E2E test page."""

    # ==================== BASIC DYNAMIC ACCESS ====================
    # A dictionary where keys come from loop items
    user_data = Variable({'user1': 'Alice', 'user2': 'Bob', 'user3': 'Charlie'})
    user_ids = Variable([{'id': 'user1'}, {'id': 'user2'}, {'id': 'user3'}])

    basic_section = Card(
        Stack(
            Text('User names accessed via dynamic key:'),
            For(
                items=user_ids,
                renderer=Stack(
                    Text(user_ids.list_item.get('id')),
                    Text(':'),
                    Text(user_data.get(user_ids.list_item.get('id'))),
                    direction='horizontal',
                    hug=True,
                ),
            ),
            hug=True,
        ),
        title='Basic Dynamic Access',
    )

    # ==================== EDITABLE DYNAMIC ACCESS ====================
    # Test that editing a dynamically-accessed nested value works
    editable_data = Variable({'item_a': 'Value A', 'item_b': 'Value B'})
    editable_items = Variable([{'key': 'item_a', 'label': 'Item A'}, {'key': 'item_b', 'label': 'Item B'}])

    editable_section = Card(
        Stack(
            Text('Editable values with dynamic keys:'),
            For(
                items=editable_items,
                renderer=Stack(
                    Text(editable_items.list_item.get('label')),
                    Text(':'),
                    Input(value=editable_data.get(editable_items.list_item.get('key'))),
                    direction='horizontal',
                    hug=True,
                ),
            ),
            Text('Current data object:'),
            Text(editable_data),
            hug=True,
        ),
        title='Editable Dynamic Access',
    )

    # ==================== DEEPLY NESTED ACCESS ====================
    # Test accessing deeply nested properties with mixed static and dynamic keys
    deep_data = Variable(
        {
            'categories': {
                'fruits': {'apple': 'Red', 'banana': 'Yellow'},
                'vegetables': {'carrot': 'Orange', 'lettuce': 'Green'},
            }
        }
    )
    category_items = Variable(
        [
            {'category': 'fruits', 'item': 'apple'},
            {'category': 'fruits', 'item': 'banana'},
            {'category': 'vegetables', 'item': 'carrot'},
            {'category': 'vegetables', 'item': 'lettuce'},
        ]
    )

    deep_section = Card(
        Stack(
            Text('Deeply nested access (static + dynamic keys):'),
            For(
                items=category_items,
                renderer=Stack(
                    Text(category_items.list_item.get('category')),
                    Text('/'),
                    Text(category_items.list_item.get('item')),
                    Text(':'),
                    Text(
                        deep_data.get('categories')
                        .get(category_items.list_item.get('category'))
                        .get(category_items.list_item.get('item'))
                    ),
                    direction='horizontal',
                    hug=True,
                ),
            ),
            hug=True,
        ),
        title='Deeply Nested Access',
    )

    # ==================== DYNAMIC LIST UPDATE ====================
    # Test that the For loop updates when the items list changes
    dynamic_data = Variable({'x': 'X Value', 'y': 'Y Value', 'z': 'Z Value'})
    dynamic_items = Variable([{'id': 'x'}])

    @action
    async def add_item(ctx: ActionCtx, items: list, data: dict):
        current_keys = {item['id'] for item in items}
        all_keys = set(data.keys())
        available = all_keys - current_keys
        if available:
            new_key = sorted(available)[0]
            await ctx.update(dynamic_items, [*items, {'id': new_key}])

    @action
    async def remove_item(ctx: ActionCtx, items: list):
        if len(items) > 0:
            await ctx.update(dynamic_items, items[:-1])

    @action
    async def reset_items(ctx: ActionCtx):
        await ctx.update(dynamic_items, [{'id': 'x'}])

    dynamic_section = Card(
        Stack(
            Stack(
                Button('Add Item', onclick=add_item(dynamic_items, dynamic_data)),
                Button('Remove Item', onclick=remove_item(dynamic_items)),
                Button('Reset', onclick=reset_items()),
                direction='horizontal',
                hug=True,
            ),
            Text('Dynamic list with dynamic key access:'),
            For(
                items=dynamic_items,
                renderer=Stack(
                    Text(dynamic_items.list_item.get('id')),
                    Text(':'),
                    Text(dynamic_data.get(dynamic_items.list_item.get('id'))),
                    direction='horizontal',
                    hug=True,
                ),
                placeholder=Text('No items'),
            ),
            hug=True,
        ),
        title='Dynamic List Update',
    )

    return Stack(basic_section, editable_section, deep_section, dynamic_section)
